### ISSUE TITLE
nonEmptyArray.head is throwing an unexpected error

### ISSUE THREAD

**Issue by AHBruns** [gcanti/fp-ts#1443]
## 🐛 Bug report

I'm using `nonEmptyArray.head` to get the head of a non-empty array. However, it seems to be throwing a runtime exception. I replaced it with `(v) => v[0]` (which should work identically, afaik), and I don't get the runtime error. So, I'm convinced it's not my code that's the problem as of now.

### Current Behavior
The following:

```typescript
const selectRandomOperation = (availableOperations: Set<Operation>) =>
  pipe(
    availableOperations,
    set.toArray(OperationOrd),
    (v: Array<Operation>): nonEmptyArray.NonEmptyArray<Operation> => {
      const neaR = nonEmptyArray.fromArray(v);
      if (option.isNone(neaR))
        throw new Error("availableOperations must not be empty");
      else return neaR.value;
    },
    // (v) => v[0],
    nonEmptyArray.head
  );
```
throws the following error at runtime:

<img width="782" alt="Screen Shot 2021-03-25 at 9 20 22 AM" src="https://user-images.githubusercontent.com/33048856/112488246-5c90f700-8d4b-11eb-924c-084368b64302.png">

However,

```typescript
const selectRandomOperation = (availableOperations: Set<Operation>) =>
  pipe(
    availableOperations,
    set.toArray(OperationOrd),
    (v: Array<Operation>): nonEmptyArray.NonEmptyArray<Operation> => {
      const neaR = nonEmptyArray.fromArray(v);
      if (option.isNone(neaR))
        throw new Error("availableOperations must not be empty");
      else return neaR.value;
    },
    (v) => v[0]
    // nonEmptyArray.head
  );
```

throws no errors. One final check was to use the `fp-ts-std` debug function `traceWithValue` to get the value just before it gets passed into `nonEmptyArray.head`. The following:

```typescript
const selectRandomOperation = (availableOperations: Set<Operation>) =>
  pipe(
    availableOperations,
    set.toArray(OperationOrd),
    (v: Array<Operation>): nonEmptyArray.NonEmptyArray<Operation> => {
      const neaR = nonEmptyArray.fromArray(v);
      if (option.isNone(neaR))
        throw new Error("availableOperations must not be empty");
      else return neaR.value;
    },
    debug.traceWithValue("value:"),
    nonEmptyArray.head
  );
```

gives me the following `console.log` and runtime exception:

<img width="495" alt="Screen Shot 2021-03-25 at 9 25 21 AM" src="https://user-images.githubusercontent.com/33048856/112489009-0c666480-8d4c-11eb-93c5-ec2fa1bb6cb4.png">

### Expected behavior

It should give me the head of the array rather than a runtime exception.

### Reproducible example

I think my current behavior section covers this, but I can try making a simpler repro if needed.

### Suggested solution(s)

I couldn't tell you. It seems that it thinks `nonEmptyArray.head` isn't a function at runtime. Maybe it's being exported wrong?

### Additional context

Nothing comes to mind.

### Your environment

I am using a `"fp-ts": "^2.10.0-rc.4"`. So, there's a possibility this is a regression introduced since the last major release.

| Software   | Version(s) |
| ---------- | ---------- |
| fp-ts           | ^2.10.0-rc.4 |
| TypeScript | 4.2.3 |

**Comment by AHBruns** [comment]
Okay, so I dug into the fp-ts codebase a bit and it seems `nonEmptyArray.head` is an alias for an alias for this function:

```typescript
export const extract: Comonad1<URI>['extract'] = (as) => as[0]
```
This looks identical to the function I replaced `nonEmptyArray.head` with. So, I'm pretty confused...

**Comment by AHBruns** [comment]
```typescript
const selectRandomOperation = (availableOperations: Set<Operation>) =>
  pipe(
    availableOperations,
    set.toArray(OperationOrd),
    (v: Array<Operation>): nonEmptyArray.NonEmptyArray<Operation> => {
      const neaR = nonEmptyArray.fromArray(v);
      if (option.isNone(neaR))
        throw new Error("availableOperations must not be empty");
      else return neaR.value;
    },

    debug.traceWithValue(`about to apply ${nonEmptyArray.head} to `),
    nonEmptyArray.head
  );
```
gives: 

<img width="418" alt="Screen Shot 2021-03-25 at 9 46 57 AM" src="https://user-images.githubusercontent.com/33048856/112492231-1178e300-8d4f-11eb-9ee7-3d3517e042e0.png">

However, VSCode intellisense seems to agree that that function exists, and my import is correct:
```typescript
import * as nonEmptyArray from "fp-ts/NonEmptyArray";
```

**Comment by gcanti** [comment]
@AHBruns looks like a circular module dependencies issue, let me dig into that

**Comment by steida** [comment]
@gcanti I am using https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-cycle.md It helps a lot.

**Comment by gcanti** [comment]
Yep, found it:

- `NonEmptyArray` -> `ReadonlyNonEmptyArray` (`import * as RNEA from './ReadonlyNonEmptyArray'`)
- `ReadonlyNonEmptyArray` -> `NonEmptyArray` (`import { NonEmptyArray, fromReadonlyNonEmptyArray } from './NonEmptyArray'`)

The culprit is the `fromReadonlyNonEmptyArray` import (the other import is `NonEmptyArray` which is a type so it doesn't matter).

@steida thanks for the suggestion, does it ignore type-only imports though? (example: `import { NonEmptyArray } from './NonEmptyArray'` is ok)

**Comment by AHBruns** [comment]
Sweet. Thanks for the quick fix!

**Comment by gcanti** [comment]
@AHBruns just released 2.10.0-rc.6 could you please try it out?

**Comment by AHBruns** [comment]
Okay, I just upgraded, and it worked as expected. Thanks again!
